doc/PROTOCOL-HTTP2.md

```diff
@@ -205,6 +205,16 @@ All GRPC calls need to specify an internal ID. We will use HTTP2 stream-ids as c
 ##### Data Frames
 DATA frame boundaries have no relation to **Length-Prefixed-Message** boundaries and implementations should make no assumptions about their alignment.
 
+##### Early Server Closure
+
+If a server responds with **Trailers** before the client's END_STREAM, then the
+server should send RST_STREAM with NO_ERROR error code immediately following
+that HEADERS frame as suggested by RFC 7540 Section 8.1 to allow the stream to
+be deallocated. Similarly, when a client receives **Trailers** before the
+client's END_STREAM, then the client should send RST_STREAM with NO_ERROR error
+code to allow the stream to be deallocted. This is done on both client and
+server because the remote behavior cannot be guaranteed.
```

Comment:
Is the client sending RST_STREAM in this situation correct? Typically the server would have sent RST_STREAM to the client if the server finishes the response without reading to the end of the request stream. A client shouldn't send an RST_STREAM in response to an RST_STREAM (to avoid recursive loop)

I think the content covered here is standard HTTP/2 behavior. Probably best to leave it to the HTTP/2 spec to avoid the gRPC spec getting it wrong.

Reply:
> Typically the server would have sent RST_STREAM to the client if the server finishes the response without reading to the end of the request stream.

We'd _hope_ that is the server's behavior, but it isn't guaranteed. Quite annoyingly, the RST_STREAM from the server is a _MAY_. RFC 7540 §8.1:

> An HTTP response is complete after the server sends — or the client receives — a frame with the END_STREAM flag set (including any CONTINUATION frames needed to complete a header block). A server can send a complete response prior to the client sending an entire request if the response does not depend on any portion of the request that has not been sent and received. When this is true, a server MAY request that the client abort transmission of a request without error by sending a RST_STREAM with an error code of NO_ERROR after sending a complete response (i.e., a frame with the END_STREAM flag).

> A client shouldn't send an RST_STREAM in response to an RST_STREAM (to avoid recursive loop)

Definitely. But that's not what I said here. I said it is response to "Trailers", which implies END_STREAM. That would be before the RST_STREAM is received/processed. It'd be _really nifty_ to avoid the RST_STREAM from the client, but there doesn't seem to be a easy way to send it only when necessary.

I agree this is really an HTTP/2 issue, but we know the expected semantics, the RFC really is not clear how it should work, and it has bit each of Java/Go/C. Even with proper grpc servers, proxies still cause problems.